op-node/rollup/sync/start.go

```diff
@@ -41,163 +28,197 @@ import (
 	"errors"
 	"fmt"
 
-	"github.com/ethereum/go-ethereum/common"
-
 	"github.com/ethereum-optimism/optimism/op-node/eth"
 	"github.com/ethereum-optimism/optimism/op-node/rollup"
+	"github.com/ethereum/go-ethereum"
+	"github.com/ethereum/go-ethereum/common"
 )
 
 type L1Chain interface {
 	L1BlockRefByLabel(ctx context.Context, label eth.BlockLabel) (eth.L1BlockRef, error)
 	L1BlockRefByNumber(ctx context.Context, number uint64) (eth.L1BlockRef, error)
+	L1BlockRefByHash(ctx context.Context, hash common.Hash) (eth.L1BlockRef, error)
 }
 
 type L2Chain interface {
 	L2BlockRefByHash(ctx context.Context, l2Hash common.Hash) (eth.L2BlockRef, error)
+	L2BlockRefByLabel(ctx context.Context, label eth.BlockLabel) (eth.L2BlockRef, error)
 }
 
+var ReorgFinalizedErr = errors.New("cannot reorg finalized block")
 var WrongChainErr = errors.New("wrong chain")
 var TooDeepReorgErr = errors.New("reorg is too deep")
 
 const MaxReorgDepth = 500
 
-// isCanonical returns the following values:
-//   - `aheadOrCanonical: true if the supplied block is ahead of the known head of the L1 chain,
-//     or canonical in the L1 chain.
-//   - `canonical`: true if the block is canonical in the L1 chain.
-func isAheadOrCanonical(ctx context.Context, l1 L1Chain, block eth.BlockID) (aheadOrCanonical bool, canonical bool, err error) {
-	if l1Head, err := l1.L1BlockRefByLabel(ctx, eth.Unsafe); err != nil {
-		return false, false, err
-	} else if block.Number > l1Head.Number {
-		return true, false, nil
-	} else if canonical, err := l1.L1BlockRefByNumber(ctx, block.Number); err != nil {
-		return false, false, err
-	} else {
-		canonical := canonical.Hash == block.Hash
-		return canonical, canonical, nil
+type FindHeadsResult struct {
+	Unsafe    eth.L2BlockRef
+	Safe      eth.L2BlockRef
+	Finalized eth.L2BlockRef
```

Comment:
we don't strictly need the `Finalized` head b/c we never actually use it.

Reply:
We do use it, we need to make sure we don't go past it, and that we use the current engine value instead of a past value in rollup node. So retrieving it when we retrieve safe/unsafe works well.